File: android/app/src/main/java/app/notomorrow/feature/progress/Milestones.kt
package app.notomorrow.feature.progress

import androidx.compose.runtime.Immutable
import app.notomorrow.data.relation.CompletedSetRow
import app.notomorrow.feature.workout.WorkoutStrings
import app.notomorrow.model.SetKind
import app.notomorrow.model.WeightUnit
import app.notomorrow.util.Fmt
import app.notomorrow.util.LocaleProvider
import app.notomorrow.util.Localizer
import app.notomorrow.util.S
import java.text.DecimalFormat
import java.text.NumberFormat
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.util.Locale

/**
 * Lifetime milestones — port of `Milestones.swift`: derived from the finished workouts alone,
 * nothing stored. Each milestone is a tier of one track — workouts completed, total volume, the
 * longest run of ISO weeks with a workout, and three lifts against body weight — and is either
 * achieved by the workout that crossed it or shows how far along it is.
 */
object Milestones {

    enum class Kind {
        Workouts, Volume, WeekStreak, Bench, Squat, Deadlift;

        val isLift: Boolean get() = this in LIFTS
    }

    /** The three lifts measured against body weight. */
    val LIFTS: List<Kind> = listOf(Kind.Bench, Kind.Squat, Kind.Deadlift)

    // MARK: - Tiers

    val WORKOUT_TIERS: List<Int> = listOf(1, 10, 25, 50, 100, 250, 500)

    /** 1 t, 10 t, 50 t, 100 t, 500 t, 1000 t — in kilograms. */
    val VOLUME_TIERS_KG: List<Double> = listOf(1_000.0, 10_000.0, 50_000.0, 100_000.0, 500_000.0, 1_000_000.0)
    val WEEK_STREAK_TIERS: List<Int> = listOf(4, 12, 26, 52)

    /** Body-weight multiple each lift is measured against. */
    fun multiplier(kind: Kind): Double? = when (kind) {
        Kind.Bench -> 1.0
        Kind.Squat -> 1.5
        Kind.Deadlift -> 2.0
        else -> null
    }

    /** The catalog exercises (free-exercise-db ids) that count as each lift: the barbell competition movements. */
    val LIFT_EXERCISE_IDS: Map<Kind, Set<String>> = mapOf(
        Kind.Bench to setOf("Barbell_Bench_Press_-_Medium_Grip", "Bench_Press_-_Powerlifting"),
        Kind.Squat to setOf("Barbell_Squat", "Barbell_Full_Squat"),
        Kind.Deadlift to setOf("Barbell_Deadlift"),
    )

    fun lift(exerciseId: String?): Kind? =
        exerciseId?.let { id -> LIFTS.firstOrNull { LIFT_EXERCISE_IDS[it]?.contains(id) == true } }

    // MARK: - Input / output

    /** One finished workout with at least one completed set. */
    data class Session(
        val workoutId: String,
        val startedAt: Long,
        /** Its start day, in the zone the weeks are counted in. */
        val day: LocalDate,
        /** Working-set volume (warm-ups left out, as `Workout.totalVolumeKg`). */
        val volumeKg: Double,
        /** Heaviest completed working set of each lift in the workout. */
        val heaviestKg: Map<Kind, Double> = emptyMap(),
    )

    @Immutable
    data class Achievement(val workoutId: String, val day: LocalDate)

    @Immutable
    data class Milestone(
        val kind: Kind,
        /** Workouts, kilograms, weeks — for a lift, the kilograms to lift ([multiplier] × body weight). */
        val target: Double,
        /** Where the track stands now, in [target]'s unit (a lift: its heaviest set). */
        val current: Double,
        val achieved: Achievement?,
    ) {
        val id: String get() = "${kind.name}-$target"
        val isAchieved: Boolean get() = achieved != null

        /** 0…1 toward [target]; 1 once achieved. */
        val progress: Double
            get() = when {
                isAchieved -> 1.0
                target <= 0 -> 0.0
                else -> (current / target).coerceIn(0.0, 1.0)
            }

        val multiplier: Double? get() = Milestones.multiplier(kind)
    }

    /**
     * One [Session] per finished workout in [rows] (completed sets only, so a workout without one
     * never appears) — as the training calendar and the weekly stats count them.
     */
    fun sessions(rows: List<CompletedSetRow>, zone: ZoneId): List<Session> =
        rows.filter { it.workoutEndedAt != null }
            .groupBy { it.workoutId }
            .map { (id, sets) ->
                val first = sets.first()
                val working = sets.filter { it.kind != SetKind.Warmup }
                val heaviest = mutableMapOf<Kind, Double>()
                for (set in working) {
                    val lift = lift(set.exerciseId) ?: continue
                    if (set.reps <= 0) continue
                    heaviest[lift] = maxOf(heaviest[lift] ?: 0.0, set.weightKg)
                }
                Session(
                    workoutId = id,
                    startedAt = first.workoutStartedAt,
                    day = Instant.ofEpochMilli(first.workoutStartedAt).atZone(zone).toLocalDate(),
                    volumeKg = working.sumOf { it.weightKg * it.reps },
                    heaviestKg = heaviest,
                )
            }

    /** The latest logged body weight, else the one in the profile; `null` without a positive one. */
    fun bodyWeight(latestEntryKg: Double?, profileKg: Double?): Double? =
        latestEntryKg?.takeIf { it > 0 } ?: profileKg?.takeIf { it > 0 }

    // MARK: - Evaluation

    /** Every tier of every track, in track order then tier order. The lifts are left out without a body weight. */
    fun evaluate(sessions: List<Session>, bodyWeightKg: Double?): List<Milestone> {
        val ordered = sessions.sortedWith(compareBy({ it.startedAt }, { it.workoutId }))
        val result = mutableListOf<Milestone>()
        result += countTrack(Kind.Workouts, WORKOUT_TIERS.map { it.toDouble() }, ordered) { index -> index + 1.0 }
        var running = 0.0
        val volumes = ordered.map { running += it.volumeKg; running }
        result += countTrack(Kind.Volume, VOLUME_TIERS_KG, ordered) { index -> volumes[index] }
        val streaks = longestStreaks(ordered.map { it.day })
        result += countTrack(Kind.WeekStreak, WEEK_STREAK_TIERS.map { it.toDouble() }, ordered) { index ->
            streaks[index].toDouble()
        }
        if (bodyWeightKg != null && bodyWeightKg > 0) {
            for (kind in LIFTS) {
                val target = (multiplier(kind) ?: continue) * bodyWeightKg
                var best = 0.0
                var achieved: Achievement? = null
                for (session in ordered) {
                    val heaviest = session.heaviestKg[kind] ?: 0.0
                    best = maxOf(best, heaviest)
                    if (achieved == null && heaviest >= target) achieved = Achievement(session.workoutId, session.day)
                }
                result += Milestone(kind, target, best, achieved)
            }
        }
        return result
    }

    /**
     * A monotonic track: `value(index)` is where the track stands after `ordered[index]`; each
     * tier is achieved by the first session whose value reaches it.
     */
    private fun countTrack(
        kind: Kind,
        tiers: List<Double>,
        ordered: List<Session>,
        value: (Int) -> Double,
    ): List<Milestone> {
        var current = 0.0
        val achieved = mutableMapOf<Double, Achievement>()
        ordered.forEachIndexed { index, session ->
            current = maxOf(current, value(index))
            for (tier in tiers) {
                if (tier !in achieved && current >= tier) achieved[tier] = Achievement(session.workoutId, session.day)
            }
        }
        return tiers.map { Milestone(kind, it, current, achieved[it]) }
    }

    /**
     * For each day (oldest first), the longest run so far of consecutive ISO weeks with a
     * workout — [TrainingCalendar.weekStreak]'s rule, counted over the whole history instead of
     * back from today.
     */
    fun longestStreaks(days: List<LocalDate>): List<Int> {
        var lastWeek: LocalDate? = null
        var run = 0
        var longest = 0
        return days.map { day ->
            val week = Fmt.startOfIsoWeek(day)
            run = when (lastWeek) {
                week -> run
                week.minusWeeks(1) -> run + 1
                else -> 1
            }
            lastWeek = week
            longest = maxOf(longest, run)
            longest
        }
    }

    // MARK: - Reading the list

    /** The milestones [workoutId] crossed, in track order. */
    fun crossed(workoutId: String, milestones: List<Milestone>): List<Milestone> =
        milestones.filter { it.achieved?.workoutId == workoutId }

    /** The most recently achieved milestone (a later tier wins a tie: same workout, bigger number). */
    fun latest(milestones: List<Milestone>): Milestone? {
        var best: Milestone? = null
        for (milestone in milestones) {
            val achieved = milestone.achieved ?: continue
            val current = best?.achieved
            if (current != null && current.day > achieved.day) continue
            best = milestone
        }
        return best
    }

    /** The next tier of each track still to reach, closest to done first (ties keep track order). */
    fun next(milestones: List<Milestone>): List<Milestone> {
        val seen = mutableSetOf<Kind>()
        return milestones
            .filter { !it.isAchieved && seen.add(it.kind) }
            .sortedByDescending { it.progress } // stable: ties keep track order
    }

    /** Kilograms shown in [unit]: the pound equivalent for lb users, as [Fmt.volume] / [Fmt.weight] convert. */
    fun displayAmount(kg: Double, unit: WeightUnit): Double =
        if (unit == WeightUnit.Kg) kg else kg * Fmt.LB_PER_KG

    // MARK: - Copy

    /** "50 workouts", "10 000 kg lifted", "12-week streak", "Squat 1.5× body weight". */
    fun title(
        milestone: Milestone,
        unit: WeightUnit,
        strings: Localizer,
        locale: Locale = LocaleProvider.current(),
    ): String = when (milestone.kind) {
        Kind.Workouts -> WorkoutStrings.workouts(milestone.target.toInt(), strings)
        Kind.Volume -> strings.string(S.milestone_volume_s, Fmt.volume(milestone.target, unit, locale = locale))
        Kind.WeekStreak -> WorkoutStrings.weekStreak(milestone.target.toInt(), strings)
        Kind.Bench -> strings.string(S.milestone_bench_s, multiple(milestone, locale))
        Kind.Squat -> strings.string(S.milestone_squat_s, multiple(milestone, locale))
        Kind.Deadlift -> strings.string(S.milestone_deadlift_s, multiple(milestone, locale))
    }

    /** "32 / 50", "7 450 / 10 000 kg", "95 / 100 kg". */
    fun progressText(milestone: Milestone, unit: WeightUnit, locale: Locale = LocaleProvider.current()): String =
        when (milestone.kind) {
            Kind.Workouts, Kind.WeekStreak -> "${milestone.current.toInt()} / ${milestone.target.toInt()}"
            Kind.Volume -> Fmt.volume(milestone.current, unit, withUnit = false, locale = locale) + " / " +
                Fmt.volume(milestone.target, unit, locale = locale)
            else -> Fmt.weight(milestone.current, unit, withUnit = false, locale = locale) + " / " +
                Fmt.weight(milestone.target, unit, locale = locale)
        }

    /** "1.5×" / "1,5×" */
    fun multiple(milestone: Milestone, locale: Locale = LocaleProvider.current()): String {
        val format = (NumberFormat.getNumberInstance(locale) as DecimalFormat).apply {
            minimumFractionDigits = 0
            maximumFractionDigits = 1
        }
        return format.format(milestone.multiplier ?: 1.0) + Fmt.TIMES
    }
}

File: NoTomorrow/Features/Progress/Milestones.swift
import Foundation

/// Lifetime milestones (Progress > Lifts "Milestones" card, `MilestonesView`, the finish screen's "New milestone"
/// line), derived from the finished workouts alone: nothing is stored. Each milestone is a tier of one track —
/// workouts completed, total volume, the longest run of ISO weeks with a workout, and three lifts against body
/// weight — and is either achieved by the workout that crossed it or shows how far along it is.
enum Milestones {

    enum Kind: String, CaseIterable, Hashable {
        case workouts, volume, weekStreak, bench, squat, deadlift

        /// The three lifts measured against body weight.
        static let lifts: [Kind] = [.bench, .squat, .deadlift]
        var isLift: Bool { Self.lifts.contains(self) }
    }

    // MARK: Tiers

    static let workoutTiers: [Int] = [1, 10, 25, 50, 100, 250, 500]
    /// 1 t, 10 t, 50 t, 100 t, 500 t, 1000 t — in kilograms.
    static let volumeTiersKg: [Double] = [1_000, 10_000, 50_000, 100_000, 500_000, 1_000_000]
    static let weekStreakTiers: [Int] = [4, 12, 26, 52]
    /// Body-weight multiple each lift is measured against.
    static func multiplier(_ kind: Kind) -> Double? {
        switch kind {
        case .bench: return 1
        case .squat: return 1.5
        case .deadlift: return 2
        default: return nil
        }
    }

    /// The catalog exercises (free-exercise-db ids) that count as each lift: the barbell competition movements.
    static let liftExerciseIDs: [Kind: Set<String>] = [
        .bench: ["Barbell_Bench_Press_-_Medium_Grip", "Bench_Press_-_Powerlifting"],
        .squat: ["Barbell_Squat", "Barbell_Full_Squat"],
        .deadlift: ["Barbell_Deadlift"],
    ]

    static func lift(forExerciseID id: String) -> Kind? {
        Kind.lifts.first { liftExerciseIDs[$0]?.contains(id) == true }
    }

    // MARK: Input / output

    /// One finished workout with at least one completed set.
    struct Session: Equatable {
        var workoutID: UUID
        var startedAt: Date
        /// Working-set volume (warm-ups left out, as `Workout.totalVolumeKg`).
        var volumeKg: Double
        /// Heaviest completed working set of each lift in the workout.
        var heaviestKg: [Kind: Double] = [:]
    }

    struct Achievement: Equatable {
        var workoutID: UUID
        var date: Date
    }

    struct Milestone: Identifiable, Equatable {
        let kind: Kind
        /// Workouts, kilograms, weeks — for a lift, the kilograms to lift (`multiplier` × body weight).
        let target: Double
        /// Where the track stands now, in `target`'s unit (a lift: its heaviest set).
        let current: Double
        let achieved: Achievement?

        var id: String { "\(kind.rawValue)-\(target)" }
        var isAchieved: Bool { achieved != nil }
        /// 0…1 toward `target`; 1 once achieved.
        var progress: Double {
            if isAchieved { return 1 }
            guard target > 0 else { return 0 }
            return min(1, max(0, current / target))
        }
        var multiplier: Double? { Milestones.multiplier(kind) }
    }

    // MARK: Evaluation

    /// Every tier of every track, in track order then tier order. The lifts are left out without a body weight.
    static func evaluate(sessions: [Session], bodyWeightKg: Double?, calendar: Calendar = .current) -> [Milestone] {
        let ordered = sessions.sorted { lhs, rhs in
            if lhs.startedAt != rhs.startedAt { return lhs.startedAt < rhs.startedAt }
            return lhs.workoutID.uuidString < rhs.workoutID.uuidString
        }
        var result: [Milestone] = []
        result += countTrack(.workouts, tiers: workoutTiers.map(Double.init), ordered: ordered) { index, _ in
            Double(index + 1)
        }
        var runningVolume = 0.0
        let volumes = ordered.map { session -> Double in
            runningVolume += session.volumeKg
            return runningVolume
        }
        result += countTrack(.volume, tiers: volumeTiersKg, ordered: ordered) { index, _ in volumes[index] }
        let streaks = longestStreaks(ordered.map(\.startedAt), calendar: calendar)
        result += countTrack(.weekStreak, tiers: weekStreakTiers.map(Double.init), ordered: ordered) { index, _ in
            Double(streaks[index])
        }
        if let bodyWeightKg, bodyWeightKg > 0 {
            for kind in Kind.lifts {
                guard let factor = Self.multiplier(kind) else { continue }
                let target = factor * bodyWeightKg
                var best = 0.0
                var achieved: Achievement?
                for session in ordered {
                    let heaviest = session.heaviestKg[kind] ?? 0
                    best = max(best, heaviest)
                    if achieved == nil, heaviest >= target {
                        achieved = Achievement(workoutID: session.workoutID, date: session.startedAt)
                    }
                }
                result.append(Milestone(kind: kind, target: target, current: best, achieved: achieved))
            }
        }
        return result
    }

    /// A monotonic track: `value(index, session)` is where the track stands after `ordered[index]`; each tier is
    /// achieved by the first session whose value reaches it.
    private static func countTrack(_ kind: Kind, tiers: [Double], ordered: [Session],
                                   value: (Int, Session) -> Double) -> [Milestone] {
        var current = 0.0
        var achieved: [Double: Achievement] = [:]
        for (index, session) in ordered.enumerated() {
            current = max(current, value(index, session))
            for tier in tiers where achieved[tier] == nil && current >= tier {
                achieved[tier] = Achievement(workoutID: session.workoutID, date: session.startedAt)
            }
        }
        return tiers.map { Milestone(kind: kind, target: $0, current: current, achieved: achieved[$0]) }
    }

    /// For each date (oldest first), the longest run so far of consecutive ISO weeks with a workout — the
    /// training calendar's `weekStreak` rule, counted over the whole history instead of back from today.
    static func longestStreaks(_ dates: [Date], calendar: Calendar = .current) -> [Int] {
        var result: [Int] = []
        var lastWeek: Date?
        var run = 0
        var longest = 0
        for date in dates {
            let week = calendar.startOfISOWeek(for: date)
            if let lastWeek, week == lastWeek {
                // Same week: the run stands.
            } else if let lastWeek, let next = calendar.date(byAdding: .day, value: 7, to: lastWeek),
                      calendar.startOfISOWeek(for: next) == week {
                run += 1
            } else {
                run = 1
            }
            lastWeek = week
            longest = max(longest, run)
            result.append(longest)
        }
        return result
    }

    // MARK: Reading the list

    /// The milestones `workoutID` crossed, in track order.
    static func crossed(by workoutID: UUID, in milestones: [Milestone]) -> [Milestone] {
        milestones.filter { $0.achieved?.workoutID == workoutID }
    }

    /// The most recently achieved milestone (a later tier wins a tie: same workout, bigger number).
    static func latest(_ milestones: [Milestone]) -> Milestone? {
        var best: Milestone?
        for milestone in milestones {
            guard let achieved = milestone.achieved else { continue }
            if let current = best?.achieved, current.date > achieved.date { continue }
            best = milestone
        }
        return best
    }

    /// The next tier of each track still to reach, closest to done first (ties keep track order).
    static func next(_ milestones: [Milestone]) -> [Milestone] {
        var seen = Set<Kind>()
        let upcoming = milestones.filter { milestone in
            guard !milestone.isAchieved, !seen.contains(milestone.kind) else { return false }
            seen.insert(milestone.kind)
            return true
        }
        return upcoming.enumerated()
            .sorted { lhs, rhs in
                if lhs.element.progress != rhs.element.progress { return lhs.element.progress > rhs.element.progress }
                return lhs.offset < rhs.offset
            }
            .map(\.element)
    }

    /// Kilograms shown in `unit`: the pound equivalent for lb users, as `Fmt.volume` / `Fmt.weight` convert.
    static func displayAmount(_ kg: Double, unit: WeightUnit) -> Double {
        unit == .kg ? kg : kg * Fmt.lbPerKg
    }
}
